code-gen/src/main/scala/scalapb/zio_grpc/ZioCodeGenerator.scala

```diff
@@ -598,33 +549,51 @@ class ZioFilePrinter(
-      val clientCall = method.streamType match {
-        case StreamType.Unary           => s"$ClientCalls.withMetadata.unaryCall"
-        case StreamType.ClientStreaming => s"$ClientCalls.withMetadata.clientStreamingCall"
-        case StreamType.ServerStreaming => s"$ClientCalls.withMetadata.serverStreamingCall"
-        case StreamType.Bidirectional   => s"$ClientCalls.withMetadata.bidiCall"
+      val Req = method.inputType.scalaType
+      val Res = method.outputType.scalaType
+
+      val clientCall      = method.streamType match {
+        case StreamType.Unary           => s"$ClientCalls.withMetadata.unaryCall[$Req, $Res]"
+        case StreamType.ClientStreaming => s"$ClientCalls.withMetadata.clientStreamingCall[$Req, $Res]"
+        case StreamType.ServerStreaming => s"$ClientCalls.withMetadata.serverStreamingCall[$Req, $Res]"
+        case StreamType.Bidirectional   => s"$ClientCalls.withMetadata.bidiCall[$Req, $Res]"
       }
-      val prefix     = method.streamType match {
-        case StreamType.Unary           => s"metadata.flatMap"
-        case StreamType.ClientStreaming => s"metadata.flatMap"
-        case StreamType.ServerStreaming =>
-          s"zio.stream.ZStream.fromZIO(metadata).flatMap"
-        case StreamType.Bidirectional   =>
-          s"zio.stream.ZStream.fromZIO(metadata).flatMap"
+      val transformMethod = method.streamType match {
+        case StreamType.Unary | StreamType.ClientStreaming         => "effect"
+        case StreamType.ServerStreaming | StreamType.Bidirectional => "stream"
+      }
+      val metadata        = method.streamType match {
+        case StreamType.Unary | StreamType.ClientStreaming         =>
+          s"context.metadata.flatMap"
+        case StreamType.ServerStreaming | StreamType.Bidirectional =>
+          s"zio.stream.ZStream.fromZIO(context.metadata).flatMap"
       }
       fp.add(
         clientWithResponseMetadataSignature(
           method,
           envOutType = "Any"
-        ) + s" = $prefix { headers => $clientCall("
+        ) + s" ="
       ).indent
-        .add(
-          s"channel, ${method.grpcDescriptor.fullName}, callOptions,"
+        .add(s"transforms.$transformMethod { context => ")
+        .indented(
+          _.add(s"$metadata { headers =>")
+            .indented(
+              _.add(
+                s"$clientCall("
+              )
+                .indented(
+                  _.add(s"channel,")
+                    .add(s"context.method.asInstanceOf[$methodDescriptor[$Req, $Res]],")
+                    .add("context.options,")
+                    .add("headers,")
+                    .add("request")
+                )
+                .add(")")
+            )
+            .add("}")
         )
-        .add(s"headers,")
-        .add(s"request")
+        .add(s"}($CallContext(${method.grpcDescriptor.fullName}, $CallOptions.DEFAULT, $SafeMetadata.make))")
```

Comment:
I think this discards the `metadata` parameter provided to the `ServiceStub`? `$SafeMetadata.make` should be just `metadata`.

Reply:
There is no more `metadata` parameter provided to the `ServiceStub`. The constructor of the `ServiceStub` now takes a `ZTransform` (which is identity by default)

Reply:
Got it! Left a few comments about API compatibility then.